### PR TITLE
Refactor of pizza crate code and various other small pizza related code changes

### PR DESCRIPTION
**Pull request by flowercuco**
## About The Pull Request

The first and biggest thing that is in this pull request is changing the old pizza crate code from something that is hard to parse and full of blahblach = C stuff into things that I believe are more in line with modern standards, which adds the potential for expansion or additions to pizza crates with more ease if, for example, someone wanted to make a flatbread crate or something. 

I've also changed all of the lizard flatbreads from just /food/pizza/ into being /food/pizza/flatbread so that I could fax them, along with sliced bread. 

Then I went in on pizzabox code adding various screentips and balloon alerts so that holding stacks of pizza and manipulating them doesnt fill your chat with stuff and finally to pizza code to add screentips.
dreamseeker_SF20VOXlNJ
dreamseeker_wtROqx0gEc
dreamseeker_2dfG6QkV4R

EDIT: I've changed the code so that instead of using the tool context screentips add element, it adds screentips via the processable add element which means all processable food now has a default screentip of "process" as a result of this PR

## Why It's Good For The Game

Old pizza crate code was unclear and now is clear so it should be easier to know what to mess with if someone wants to change them further or add or expand them. 
We all like screentips!
Old pizzaboxes had way too many unnecessary messages getting printed to chat.
You can fax root flatbread but not flatbread with toppings on them and while that can make a kind of sense you can fax pizza slices with toppings on them! Bread slices just also makes sense to add while I'm at it.

## Changelog

:cl:
add: you can fax flatbreads and slices of bread now, changed all flatbreads into children of pizza/flatbread to allow for this
qol: adds screentips to various pizzabox action (stacking, when you can take pizzas out, pizza box bomb stuff)
qol: adds a way to add screentips directly into processable food via the add element processable proc
refactor: pizza crates work the same but more clearly and more variable
/:cl:

### CODE REVIEW

**Comment by dalmationer** [comment]
A sol's crate!?

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/game/objects/items/food/pizza.dm [hunk 1]
```diff
@@ -390,3 +404,12 @@
...
 	foodtypes = TOXIC
+
+///screentips for pizza
+/obj/item/food/pizza/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+
+	if(held_item?.tool_behaviour == TOOL_KNIFE || TOOL_SCALPEL || TOOL_SAW)
+		if(slice_type == null)
+			return
+		context[SCREENTIP_CONTEXT_LMB] = "Slice pizza"
+		return CONTEXTUAL_SCREENTIP_SET
```
[the remark this answers is not in the document]
Comment (flowercuco) [reply code/game/objects/items/food/pizza.dm:415]: i will look into the feasibility of (me personally) trying to add contextual screentips to the element because your solution wrt tools is what i was hoping i would not have to do haha. 

thank you!

[the remark this answers is not in the document]
Comment (flowercuco) [reply code/game/objects/items/food/pizza.dm]: it turns out that someone WAS me! and only with medium help from @Jacquerel !!!

File: code/datums/elements/food/processable.dm [hunk 2]
```diff
@@ -54,3 +62,22 @@
...
 	else
 		examine_list += span_notice("It can be turned into \a [initial(result_atom_type.name)] with <b>[tool_behaviour_name(tool_behaviour)]</b>!")
+
+///Adds context sensitivy directly to the processable file for screentips
+/datum/element/processable/proc/on_requesting_context_from_item(
+	datum/source,
+	list/context,
+	obj/item/held_item,
+	mob/user,
+)
```
Comment (san7890) [inline code/datums/elements/food/processable.dm:72]:
```suggestion
/datum/element/processable/proc/on_requesting_context_from_item(datum/source, list/context, obj/item/held_item, mob/user)
```

you don't have THAT many arguments. also, might be worthwhile to DMdoc what each of these mean/do

File: code/modules/cargo/packs/organic.dm [hunk 3]
```diff
@@ -185,52 +181,58 @@
...
+/datum/supply_pack/organic/pizza/proc/add_anomalous(obj/structure/closet/crate/new_crate)
+	if(anomalous_box_provided)
+		return FALSE
+	if(!prob(infinite_pizza_chance))
+		return FALSE
+	new /obj/item/pizzabox/infinite(new_crate)
+	anomalous_box_provided = TRUE
+	log_game("An anomalous pizza box was provided in a pizza crate at during cargo delivery.")
+	if(prob(50))
+		addtimer(CALLBACK(src, .proc/anomalous_pizza_report), rand(300, 1800))
```
Comment (san7890) [inline code/modules/cargo/packs/organic.dm:206]:
```suggestion
		addtimer(CALLBACK(src, .proc/anomalous_pizza_report), rand(30 SECONDS, 180 SECONDS))
```
labelling units helps people know the exact intent, may as well touch it up while you're doing stuff ehre

File: code/datums/elements/food/processable.dm [hunk 4]
```diff
@@ -12,8 +12,11 @@
 	var/amount_created
 	///Whether or not the atom being processed has to be on a table or tray to process it
 	var/table_required
+	///Verb used in processing food (such as slice, flatten), defaults to process
+	var/screentip_verb
 
-/datum/element/processable/Attach(datum/target, tool_behaviour, result_atom_type, amount_created = 3, time_to_process = 2 SECONDS, table_required = FALSE)
+/datum/element/processable/Attach(datum/target, tool_behaviour, result_atom_type,
+	amount_created = 3, time_to_process = 2 SECONDS, table_required = FALSE, screentip_verb = "Process")
```
Comment (san7890) [inline code/datums/elements/food/processable.dm:19]:
```suggestion
/datum/element/processable/Attach(datum/target, tool_behaviour, result_atom_type, amount_created = 3, time_to_process = 2 SECONDS, table_required = FALSE, screentip_verb = "Process")
```

it just looks odd otherwise imo

File: code/datums/elements/food/processable.dm (hunk 2 above)
Comment (flowercuco) [reply code/datums/elements/food/processable.dm]: let me know if my documentation is off or wrong, but i think that theyre correct

File: code/datums/elements/food/processable.dm [hunk 5]
```diff
@@ -54,3 +61,24 @@
...
+
+/**
+ * Adds context sensitivy directly to the processable file for screentips
+ * Arguments:
+ * * source - refers to item that will display its screentip
+ * * context - refers to, in this case, an item that can be proccessed into another item via add element proccessable
+ * * held_item - refers to tool used by add element proccessable to process one item into another item
+ * * user - refers to user who will see the screentip when the proper context and tool are there
+ */
+/datum/element/processable/proc/on_requesting_context_from_item(datum/source, list/context, obj/item/held_item, mob/user,)
```
Comment (san7890) [inline code/datums/elements/food/processable.dm:73]:
```suggestion
/datum/element/processable/proc/on_requesting_context_from_item(datum/source, list/context, obj/item/held_item, mob/user)
```

Cruft when you flattened it into one line. won't affect compile but it looks bad

File: code/datums/elements/food/processable.dm (hunk 2 above)
Comment (san7890) [reply code/datums/elements/food/processable.dm]: The format looks alright to my eyes

File: code/modules/cargo/packs/organic.dm [hunk 6]
```diff
@@ -185,52 +181,58 @@
...
-			P.update_appearance()
+/// adds a randomized pizza from the pizza list
+/datum/supply_pack/organic/pizza/proc/add_normal_pizza(obj/structure/closet/crate/new_crate, list/rng_pizza_list)
+	var/randomize_pizza = pick_weight(rng_pizza_list)
+	rng_pizza_list -= randomize_pizza
+	var/obj/item/pizzabox/new_pizza_box = new(new_crate)
+	new_pizza_box.pizza = new randomize_pizza
+	new_pizza_box.boxtag = new_pizza_box.pizza.boxtag
+	new_pizza_box.boxtag_set = TRUE
+	new_pizza_box.update_appearance()
```
Comment (Mothblocks) [inline code/modules/cargo/packs/organic.dm:233]: What actually needs to update here?

File: code/modules/food_and_drinks/pizzabox.dm [hunk 7]
```diff
@@ -117,35 +117,38 @@
...
-		var/obj/item/stack/sheet/cardboard/cardboard = new /obj/item/stack/sheet/cardboard(user.drop_location())
-		to_chat(user, span_notice("You fold [src] into [cardboard]."))
-		user.put_in_active_hand(cardboard)
-		qdel(src)
-		return
 	update_appearance()
 
+/obj/item/pizzabox/attack_self_secondary(mob/user)
+	if(length(boxes) > 0)
+		return
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:131]:
```suggestion
		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

File: code/modules/food_and_drinks/pizzabox.dm [hunk 8]
```diff
@@ -117,35 +117,38 @@
...
+	if(pizza || bomb)
+		return
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:133]:
```suggestion
		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

File: code/modules/food_and_drinks/pizzabox.dm [hunk 9]
```diff
@@ -117,35 +117,38 @@
...
+	var/obj/item/stack/sheet/cardboard/cardboard = new /obj/item/stack/sheet/cardboard(user.drop_location())
+	balloon_alert(user, "deconstructed")
+	user.put_in_active_hand(cardboard)
+	qdel(src)
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:137]:
```suggestion
	qdel(src)
	return SECONDARY_ATTACK_CONTINUE_CHAIN
```

File: code/modules/food_and_drinks/pizzabox.dm [hunk 10]
```diff
@@ -117,35 +117,38 @@
...
-		qdel(src)
-		return
 	update_appearance()
 
+/obj/item/pizzabox/attack_self_secondary(mob/user)
+	if(length(boxes) > 0)
+		return
+	if(pizza || bomb)
+		return
+	var/obj/item/stack/sheet/cardboard/cardboard = new /obj/item/stack/sheet/cardboard(user.drop_location())
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:134]:
```suggestion
	var/obj/item/stack/sheet/cardboard/cardboard = new(user.drop_location())
```

File: code/modules/food_and_drinks/pizzabox.dm [hunk 11]
```diff
@@ -117,35 +117,38 @@
...
+	balloon_alert(user, "deconstructed")
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:135]:
```suggestion
```

This isn't necessary because there is already visual feedback where the user is looking (and the user is not where they are looking anyway)

File: code/modules/food_and_drinks/pizzabox.dm [hunk 12]
```diff
@@ -178,28 +180,26 @@
...
-			to_chat(user, span_notice("You put [newbox] on top of [src]!"))
 			newbox.update_appearance()
 			update_appearance()
 			user.regenerate_icons()
-			if(boxes.len >= 5)
-				if(prob(10 * boxes.len))
-					to_chat(user, span_danger("You can't keep holding the stack!"))
+			if(length(boxes) >= 5)
+				if(prob(10 * length(boxes)))
+					balloon_alert(user, "dropped it!")
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:188]:
```suggestion
					balloon_alert(user, "oops!")
```

nit, i just think this reads weird

File: code/modules/food_and_drinks/pizzabox.dm [hunk 13]
```diff
@@ -389,3 +389,51 @@
 	boxtag_set = FALSE
 	update_appearance() //update our boxtag to match our new pizza
 	pizza.foodtypes = nommer.dna.species.liked_food //it's our favorite!
+
+///screentips for pizzaboxes
+/obj/item/pizzabox/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	if(!held_item)
+		if(user.get_inactive_held_item() != src)
+			return
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:397]:
```suggestion
			return NONE
```

Means the same but explicitness is good, here and others

File: code/modules/food_and_drinks/pizzabox.dm [hunk 14]
```diff
@@ -389,3 +389,51 @@
...
+		if(open)
+			if(pizza)
+				context [SCREENTIP_CONTEXT_LMB] = "Remove pizza"
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/pizzabox.dm:400]: Remove the spaces here and everywhere else in between context and the [].

**Review by Mothblocks — APPROVED** [review]

File: code/modules/cargo/packs/organic.dm [hunk 15]
```diff
@@ -185,52 +181,58 @@
...
-					message_admins("An anomalous pizza box was silently created with no command report in a pizza crate delivery.")
-				continue
+/datum/supply_pack/organic/pizza/fill(obj/structure/closet/crate/new_crate)
+	. = ..()
+	var/list/rng_pizza_list = pizza_types.Copy()
+	for(var/i in 1 to 5)
+		if(add_anomalous(new_crate))
+			continue
+		if(add_boombox(new_crate))
+			continue
```
[the remark this answers is not in the document]
Comment (flowercuco) [reply code/modules/cargo/packs/organic.dm:193]: yeah this is how it worked before, it was just less clear about it

File: code/modules/cargo/packs/organic.dm (hunk 6 above)
Comment (flowercuco) [reply code/modules/cargo/packs/organic.dm:233]: the sprite doesnt start with the little text on it, so after it has a pizza in it and it knows what to write on the box it needs to update into having writing on the box

File: code/modules/food_and_drinks/pizzabox.dm (hunk 12 above)
Comment (flowercuco) [reply code/modules/food_and_drinks/pizzabox.dm:188]: i had no idea what to put in this one and "oops" was the simplest and best option the whole time 😔

File: code/modules/cargo/packs/organic.dm (hunk 6 above)
Comment (flowercuco) [reply code/modules/cargo/packs/organic.dm]: i changed it to update the icon and the description because otherwise it has the sprite with writing but doesnt tell you what the writing is.

File: code/modules/cargo/packs/organic.dm [hunk 16]
```diff
@@ -155,25 +155,21 @@
...
+	contains = list()
 	crate_name = "pizza crate"
+
 	///Whether we've provided an infinite pizza box already this shift or not.
-	var/static/anomalous_box_provided = FALSE
-	///The percentage chance (per pizza) of this supply pack to spawn an anomalous pizza box.
-	var/anna_molly_box_chance = 1
-	///Total tickets in our figurative lottery (per pizza) to decide if we create a bomb box, and if so what type. 1 to 3 create a bomb. The rest do nothing.
-	var/boombox_tickets = 100
+	var/anomalous_box_provided = FALSE
```
Comment (JohnFulpWillard) [inline code/modules/cargo/packs/organic.dm:162]: Why is this not static anymore?

File: code/modules/cargo/packs/organic.dm [hunk 17]
```diff
@@ -185,52 +181,58 @@
...
+/datum/supply_pack/organic/pizza/proc/add_normal_pizza(obj/structure/closet/crate/new_crate, list/rng_pizza_list)
+	var/randomize_pizza = pick_n_take(rng_pizza_list)
+	rng_pizza_list -= randomize_pizza
+	var/obj/item/pizzabox/new_pizza_box = new(new_crate)
+	new_pizza_box.pizza = new randomize_pizza
+	new_pizza_box.boxtag = new_pizza_box.pizza.boxtag
+	new_pizza_box.boxtag_set = TRUE
+	new_pizza_box.update_appearance(UPDATE_ICON | UPDATE_DESC)
 
+/// tells crew that an infinite pizza box exists, half of the time, based ona roll in the anamolous box proc
```
Comment (JohnFulpWillard) [inline code/modules/cargo/packs/organic.dm:235]:
```suggestion
/// tells crew that an infinite pizza box exists, half of the time, based on a roll in the anamolous box proc.
```

File: code/modules/food_and_drinks/pizzabox.dm [hunk 18]
```diff
@@ -117,35 +119,38 @@
...
 		if(pizza)
 			user.put_in_hands(pizza)
-			to_chat(user, span_notice("You take [pizza] out of [src]."))
 			pizza = null
 			update_appearance()
 		else if(bomb)
 			if(wires.is_all_cut() && bomb_defused)
 				user.put_in_hands(bomb)
-				to_chat(user, span_notice("You carefully remove the [bomb] from [src]."))
+				balloon_alert(user, "removed bomb")
```
Comment (JohnFulpWillard) [inline code/modules/food_and_drinks/pizzabox.dm:153]: this would balloon on the box, so if it's in their hand, it would balloon in the icon in their UI.
Though it might not be a problem since that's likely where the user would be looking, but if that's not what you want, you'll have to change it to ``user.balloon_alert``

File: code/modules/food_and_drinks/pizzabox.dm [hunk 19]
```diff
@@ -208,28 +210,28 @@
...
+			balloon_alert(user, "writing box tag...")
 			boxtag_set = TRUE
 			update_appearance()
 			return
 	else if(is_wire_tool(I))
 		if(wires && bomb)
 			wires.interact(user)
 	else if(istype(I, /obj/item/reagent_containers/cup))
-		to_chat(user, span_warning("That's not a pizza!"))
+		balloon_alert(user, "not a pizza!")
```
Comment (JohnFulpWillard) [inline code/modules/food_and_drinks/pizzabox.dm:234]: I think this code is here because pizza used to be a subtype of reagent containers. You should remove it honestly.

File: code/modules/food_and_drinks/pizzabox.dm [hunk 20]
```diff
@@ -178,28 +182,26 @@
...
-			to_chat(user, span_notice("You put [newbox] on top of [src]!"))
 			newbox.update_appearance()
 			update_appearance()
 			user.regenerate_icons()
-			if(boxes.len >= 5)
-				if(prob(10 * boxes.len))
-					to_chat(user, span_danger("You can't keep holding the stack!"))
+			if(length(boxes) >= 5)
+				if(prob(10 * length(boxes)))
+					balloon_alert(user, "oops!")
```
Comment (JohnFulpWillard) [inline code/modules/food_and_drinks/pizzabox.dm:190]: This is a weird balloon alert, I think it would work better as either a message to chat still, or a balloon alert to all.

File: code/modules/cargo/packs/organic.dm (hunk 16 above)
Comment (flowercuco) [reply code/modules/cargo/packs/organic.dm:162]: it doesnt need to be, the supply console only makes the datum once

File: code/datums/elements/food/processable.dm [hunk 21]
```diff
@@ -1,4 +1,4 @@
-// If an item has the processable item, it can be processed into another item with a specific tool. This adds generic behavior for those actions to make it easier to set-up generically.
+// If an item has the processable item, it can be processed into another item with a specific tool. This adds generic behaviour for those actions to make it easier to set-up generically.
```
[the remark this answers is not in the document]
Comment (flowercuco) [reply code/datums/elements/food/processable.dm:1]: there isnt a standard on english v american spelling so im going with the brits because thats how it it is called by every other variable on this document

File: code/modules/food_and_drinks/pizzabox.dm (hunk 18 above)
Comment (flowercuco) [reply code/modules/food_and_drinks/pizzabox.dm:153]: ive just tested it again, the balloon alert just comes out of the guy

image

File: code/modules/food_and_drinks/pizzabox.dm (hunk 20 above)
Comment (flowercuco) [reply code/modules/food_and_drinks/pizzabox.dm:190]: changed it to a to viewers

File: code/modules/food_and_drinks/pizzabox.dm (hunk 18 above)
Comment (Mothblocks) [reply code/modules/food_and_drinks/pizzabox.dm:153]: It uses the turf of whatever it's on, it only shows in hand if it's not equipped...I think? I forget when exactly it does but bla
